metamorph/src/main/java/org/metafacture/metamorph/Metamorph.java

```diff
@@ -215,7 +219,10 @@ public void setErrorHandler(final MorphErrorHandler errorHandler) {
     }
 
     protected void registerNamedValueReceiver(final String source, final NamedValueReceiver data) {
-        if (ELSE_KEYWORD.equals(source)) {
+        if (ELSE_AND_PASS_ENTITY_EVENTS_KEYWORD.equals(source)) {
+            this.passEntityEvents = true;
```

Comment:
The `_else` is, as I understand it, is valid for the whole stream, not bounded to some `NamedValueReceiver`. It just emits all the data that is not treated explicitly in any way in the morph.

Reply:
But it _may_ occur repeatedly¹:

```java
@Test
public void multipleElseSources() {
    metamorph = InlineMorph.in(this)
        .with("<rules>")
        .with("    <data source='_else'/>")
        .with("    <data source='_else'/>")
        .with("</rules>")
        .createConnectedTo(receiver);

    metamorph.startRecord("1");
    metamorph.literal("lit", "a");
    metamorph.startEntity("clone");
    metamorph.literal("id", "0");
    metamorph.endEntity();
    metamorph.startEntity("clone");
    metamorph.literal("id", "1");
    metamorph.endEntity();
    metamorph.endRecord();

    final InOrder ordered = inOrder(receiver);
    ordered.verify(receiver).startRecord("1");
    ordered.verify(receiver, times(2)).literal("lit", "a");
    ordered.verify(receiver, times(2)).literal("clone.id", "0");
    ordered.verify(receiver, times(2)).literal("clone.id", "1");
    ordered.verify(receiver).endRecord();
    ordered.verifyNoMoreInteractions();
    verifyNoMoreInteractions(receiver);
}
```

So this shows the wrong behaviour:

```java
@Test
public void multipleMixedElseSourcesWrong() {
    metamorph = InlineMorph.in(this)
        .with("<rules>")
        .with("    <data source='_else'/>")
        .with("    <data source='_elseAndPassEntityEvents'/>")
        .with("</rules>")
        .createConnectedTo(receiver);

    metamorph.startRecord("1");
    metamorph.literal("lit", "a");
    metamorph.startEntity("clone");
    metamorph.literal("id", "0");
    metamorph.endEntity();
    metamorph.startEntity("clone");
    metamorph.literal("id", "1");
    metamorph.endEntity();
    metamorph.endRecord();

    final InOrder ordered = inOrder(receiver);
    ordered.verify(receiver).startRecord("1");
    ordered.verify(receiver, times(2)).literal("lit", "a");
    ordered.verify(receiver).startEntity("clone");
    ordered.verify(receiver).literal("id", "0");
    ordered.verify(receiver).endEntity();
    ordered.verify(receiver).startEntity("clone");
    ordered.verify(receiver).literal("id", "0");
    ordered.verify(receiver).endEntity();
    ordered.verify(receiver).startEntity("clone");
    ordered.verify(receiver).literal("id", "1");
    ordered.verify(receiver).endEntity();
    ordered.verify(receiver).startEntity("clone");
    ordered.verify(receiver).literal("id", "1");
    ordered.verify(receiver).endEntity();
    ordered.verify(receiver).endRecord();
    ordered.verifyNoMoreInteractions();
    verifyNoMoreInteractions(receiver);
}
```

It should actually look like this to behave correctly:

```java
@Test
public void multipleMixedElseSources() {
    metamorph = InlineMorph.in(this)
        .with("<rules>")
        .with("    <data source='_else'/>")
        .with("    <data source='_elseAndPassEntityEvents'/>")
        .with("</rules>")
        .createConnectedTo(receiver);

    metamorph.startRecord("1");
    metamorph.literal("lit", "a");
    metamorph.startEntity("clone");
    metamorph.literal("id", "0");
    metamorph.endEntity();
    metamorph.startEntity("clone");
    metamorph.literal("id", "1");
    metamorph.endEntity();
    metamorph.endRecord();

    final InOrder ordered = inOrder(receiver);
    ordered.verify(receiver).startRecord("1");
    ordered.verify(receiver, times(2)).literal("lit", "a");
    ordered.verify(receiver).literal("clone.id", "0");
    ordered.verify(receiver).startEntity("clone");
    ordered.verify(receiver).literal("id", "0");
    ordered.verify(receiver).endEntity();
    ordered.verify(receiver).literal("clone.id", "1");
    ordered.verify(receiver).startEntity("clone");
    ordered.verify(receiver).literal("id", "1");
    ordered.verify(receiver).endEntity();
    ordered.verify(receiver).endRecord();
    ordered.verifyNoMoreInteractions();
    verifyNoMoreInteractions(receiver);
}
```

¹ The reasons are unclear; the original [commit](https://sourceforge.net/p/culturegraph/code/884/) is just a dump of "new developments".

Reply:
You seem correct, but I would like to not follow this here (maybe in another issue?).
I would argue that:

1. the way of expected behaviour is  determined by the tests (and these don't test the use of multiple `_else`)
2. a) it was forgotten to test b) it doesn't make sense, so it wasn't the intention, so it was not tested

My impression is that of 2.b). It could make sense, however, to ignore in Metamorph the usage of multiple `_else*` (I really find it hard to imagine a scenario where you want to have the _exact same data structure multiple times_. But if someone comes up with that desiderat, one could open an issue and implement this).

Reply:
As long as it's a conscious decision, I guess...

> I really find it hard to imagine a scenario where you want to have the _exact same data structure multiple times_

So do I, but the original author must have thought differently. What's the point of introducing `elseSource` (later [renamed](https://sourceforge.net/p/culturegraph/code/1292/) to `elseSources` - plural!) as a `List` when it could have been a simple scalar instead? (There's also the option to apply functions, though, so it's not necessarily the _exact same data_.)

My personal preferences would be (in descending order):

1. Implement the new feature in a compatible way and add tests for both the old and the new behaviour.
2. Explicitly disallow the old behaviour (single `elseSource`, warn if already set).
3. Implicitly discourage the old behaviour (your proposal).

But I defer to your judgement. I certainly agree that it would be unusual either way.

Reply:
So, let's go then with `2.` and wait for somebody opening an issue if `1.` is needed?